File: frontend/src/utils/storage-migration.test.ts
import { beforeEach, describe, expect, it } from 'vitest'

import { currentKeyFor, migrateLegacyKeys } from './storage-migration'

function fill(storage: Storage, entries: Record<string, string>) {
  for (const [key, value] of Object.entries(entries)) {
    storage.setItem(key, value)
  }
}

beforeEach(() => {
  localStorage.clear()
  sessionStorage.clear()
})

describe('legacy storage keys', () => {
  it('maps every pre-rename prefix to its current key', () => {
    expect(currentKeyFor('math-notebook:token')).toBe('questrace:token')
    expect(currentKeyFor('math-notebook:list-view')).toBe('questrace:list-view')
    expect(currentKeyFor('notebook:appearance:v1')).toBe('questrace:appearance:v1')
    expect(currentKeyFor('questrace:token')).toBeNull()
    expect(currentKeyFor('unrelated')).toBeNull()
  })

  it('copies authentication and list preferences without deleting the old keys', () => {
    fill(localStorage, {
      'math-notebook:token': 'secret-token',
      'math-notebook:auth-user': '{"user_id":1,"username":"owner"}',
      'math-notebook:list-view': 'table',
    })
    expect(migrateLegacyKeys(localStorage).sort()).toEqual([
      'questrace:auth-user',
      'questrace:list-view',
      'questrace:token',
    ])
    expect(localStorage.getItem('questrace:token')).toBe('secret-token')
    expect(localStorage.getItem('questrace:auth-user')).toBe('{"user_id":1,"username":"owner"}')
    expect(localStorage.getItem('questrace:list-view')).toBe('table')
    expect(localStorage.getItem('math-notebook:token')).toBe('secret-token')
  })

  it('copies appearance preferences including the legacy appearance name', () => {
    localStorage.setItem('notebook:appearance:v1', '{"version":1,"preset":"paper"}')
    migrateLegacyKeys(localStorage)
    expect(localStorage.getItem('questrace:appearance:v1')).toBe('{"version":1,"preset":"paper"}')
    expect(localStorage.getItem('notebook:appearance:v1')).toBe('{"version":1,"preset":"paper"}')
  })

  it('copies drafts, including the discarded-analysis snapshot, from session storage', () => {
    fill(sessionStorage, {
      'math-notebook:draft': '{"chapter":"原章节"}',
      'math-notebook:draft:before-analysis': '{"chapter":"AI章节"}',
    })
    migrateLegacyKeys(sessionStorage)
    expect(sessionStorage.getItem('questrace:draft')).toBe('{"chapter":"原章节"}')
    expect(sessionStorage.getItem('questrace:draft:before-analysis')).toBe('{"chapter":"AI章节"}')
  })

  it('never overwrites a value already stored under the current key', () => {
    fill(localStorage, {
      'math-notebook:token': 'stale-token',
      'questrace:token': 'current-token',
    })
    expect(migrateLegacyKeys(localStorage)).toEqual([])
    expect(localStorage.getItem('questrace:token')).toBe('current-token')
  })
})

File: mobile/flutter_app/test/core/storage/storage_migration_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:questrace_flutter/core/storage/storage_keys.dart';
import 'package:questrace_flutter/core/storage/storage_migration.dart';
import 'package:shared_preferences/shared_preferences.dart';

void main() {
  test('copies pre-rename keys without deleting them', () async {
    SharedPreferences.setMockInitialValues({
      'math-notebook:auth-session': '{"token":"legacy"}',
      'math-notebook:question-draft': '{"chapter":"原章节"}',
      'math-notebook:api-base-url': 'http://computer:8080',
      'math-notebook:theme-color-seed': 0xff123456,
    });
    final preferences = await SharedPreferences.getInstance();

    expect((await migrateLegacyStorage(preferences)).toSet(), {
      StorageKeys.authSession,
      StorageKeys.questionDraft,
      StorageKeys.apiBaseUrl,
      StorageKeys.themeColorSeed,
    });
    expect(preferences.getString(StorageKeys.authSession), '{"token":"legacy"}');
    expect(preferences.getString(StorageKeys.questionDraft), '{"chapter":"原章节"}');
    expect(preferences.getString(StorageKeys.apiBaseUrl), 'http://computer:8080');
    expect(preferences.getInt(StorageKeys.themeColorSeed), 0xff123456);

    // Rolling back still finds the values under their original keys.
    expect(preferences.getString('math-notebook:auth-session'), '{"token":"legacy"}');
    expect(preferences.getInt('math-notebook:theme-color-seed'), 0xff123456);
  });

  test('copies appearance preferences from the legacy appearance key', () async {
    SharedPreferences.setMockInitialValues(
        {'notebook:appearance:v1': '{"preset":"paper"}'});
    final preferences = await SharedPreferences.getInstance();

    expect(await migrateLegacyStorage(preferences),
        ['questrace:appearance:v1']);
    expect(preferences.getString('questrace:appearance:v1'), '{"preset":"paper"}');
    expect(preferences.getString('notebook:appearance:v1'), '{"preset":"paper"}');
  });

  test('never overwrites a value already stored under the current key',
      () async {
    SharedPreferences.setMockInitialValues({
      'math-notebook:api-base-url': 'http://legacy:8080',
      StorageKeys.apiBaseUrl: 'http://current:8080',
    });
    final preferences = await SharedPreferences.getInstance();

    expect(await migrateLegacyStorage(preferences), isEmpty);
    expect(preferences.getString(StorageKeys.apiBaseUrl), 'http://current:8080');
  });

  test('leaves unrelated keys untouched', () async {
    SharedPreferences.setMockInitialValues({'unrelated': 'value'});
    final preferences = await SharedPreferences.getInstance();

    expect(await migrateLegacyStorage(preferences), isEmpty);
    expect(preferences.getString('unrelated'), 'value');
  });

  test('is idempotent', () async {
    SharedPreferences.setMockInitialValues(
        {'math-notebook:auth-session': '{"token":"legacy"}'});
    final preferences = await SharedPreferences.getInstance();

    expect(await migrateLegacyStorage(preferences), [StorageKeys.authSession]);
    expect(await migrateLegacyStorage(preferences), isEmpty);
    expect(preferences.getString(StorageKeys.authSession), '{"token":"legacy"}');
  });
}
